pkg/storage/ingest/reader_test.go

```diff
@@ -0,0 +1,328 @@
+		resp.Default()
+		resp.Topics = []kmsg.OffsetCommitResponseTopic{
+			{
+				Topic:      topicName,
+				Partitions: []kmsg.OffsetCommitResponseTopicPartition{{Partition: partitionID}},
+			},
+		}
+
+		return resp, nil, true
+	})
+
+	cluster.ControlKey(kmsg.OffsetFetch.Int16(), func(request kmsg.Request) (kmsg.Response, error, bool) {
+		cluster.KeepControl()
+		commitR := request.(*kmsg.OffsetFetchRequest)
+		assert.Len(t, commitR.Groups, 1, "test only has support for one consumer group per request")
+		assert.Equal(t, commitR.Groups[0].Group, consumerGroup)
+
+		const allPartitions = -1
+		var partitionID int32
+
+		if len(commitR.Groups[0].Topics) == 0 {
+			// An empty request means fetch all topic-partitions for this group.
+			partitionID = allPartitions
+		} else {
+			partitionID = commitR.Groups[0].Topics[0].Partitions[0]
+			assert.Len(t, commitR.Groups[0], 1, "test only has support for one partition per request")
+			assert.Len(t, commitR.Groups[0].Topics[0].Partitions, 1, "test only has support for one partition per request")
+		}
+
+		var partitionsResp []kmsg.OffsetFetchResponseGroupTopicPartition
+		if partitionID == allPartitions {
+			for i := int32(1); i < numPartitions+1; i++ {
+				partitionsResp = append(partitionsResp, kmsg.OffsetFetchResponseGroupTopicPartition{
+					Partition: i,
+					Offset:    committedOffsets[i],
+				})
+			}
+		} else {
+			partitionsResp = append(partitionsResp, kmsg.OffsetFetchResponseGroupTopicPartition{
+				Partition: partitionID,
+				Offset:    committedOffsets[partitionID],
+			})
+		}
+
+		resp := request.ResponseKind().(*kmsg.OffsetFetchResponse)
+		resp.Default()
+		resp.Groups = []kmsg.OffsetFetchResponseGroup{
+			{
+				Group: consumerGroup,
+				Topics: []kmsg.OffsetFetchResponseGroupTopic{
+					{
+						Topic:      topicName,
+						Partitions: partitionsResp,
+					},
+				},
+			},
+		}
+		return resp, nil, true
+	})
+}
+
+type testConsumer struct {
+	messages chan []byte
+}
+
+func newTestConsumer(capacity int) testConsumer {
+	return testConsumer{
+		messages: make(chan []byte, capacity),
+	}
+}
+
+func (t testConsumer) consume(_ context.Context, records []record) error {
+	for _, r := range records {
+		t.messages <- r.content
+	}
+	return nil
+}
+
+// waitRecords expects to receive numRecords records within waitTimeout.
+// waitRecords waits for drainPeriod after receiving numRecords records to ensure that no more records are received.
+// waitRecords returns an error if a different number of records is received.
+func (t testConsumer) waitRecords(numRecords int, waitTimeout, drainPeriod time.Duration) ([][]byte, error) {
+	var messages [][]byte
+	timeout := time.After(waitTimeout)
+	for {
+		select {
+		case msg := <-t.messages:
+			messages = append(messages, msg)
+			if len(messages) != numRecords {
+				continue
+			}
+			if drainPeriod == 0 {
+				return messages, nil
+			}
+			timeout = time.After(drainPeriod)
```

Comment:
I don't fully understand the purpose of the drainPeriod. Looks like you can achieve the same just setting `timeout := time.After(waitPeriod + drainPeriod)`. Reading the function comment I initially thought you don't want to read from `t.messages` during the drainPeriod, but in practice it does read it from it even during the drainPeriod so I'm not sure I understand why having two timeouts.

Reply:
After reading the tests using it, now I do understand the difference. `drainPeriod` starts once the expected number of messages is received.